cpp/src/arrow/util/bitmap_generate.h

```diff
@@ -83,17 +83,19 @@ void GenerateBitsUnrolled(uint8_t* bitmap, int64_t start_offset, int64_t length,
   }
 
   int64_t remaining_bytes = remaining / 8;
+  uint8_t out_results[8];
   while (remaining_bytes-- > 0) {
-    current_byte = 0;
-    current_byte = g() ? current_byte | 0x01 : current_byte;
-    current_byte = g() ? current_byte | 0x02 : current_byte;
-    current_byte = g() ? current_byte | 0x04 : current_byte;
-    current_byte = g() ? current_byte | 0x08 : current_byte;
-    current_byte = g() ? current_byte | 0x10 : current_byte;
-    current_byte = g() ? current_byte | 0x20 : current_byte;
-    current_byte = g() ? current_byte | 0x40 : current_byte;
-    current_byte = g() ? current_byte | 0x80 : current_byte;
-    *cur++ = current_byte;
+    out_results[0] = g();
+    out_results[1] = g();
+    out_results[2] = g();
+    out_results[3] = g();
+    out_results[4] = g();
+    out_results[5] = g();
+    out_results[6] = g();
+    out_results[7] = g();
```

Comment:
This presumes that `g()` either returns a bool or 0/1 -- might want to enforce this (at compile time) but wasn't sure the best way to do that

Reply:
Maybe `out_results[0] = !!g();` ?

Reply:
I'm adding a compile-time assertion